python_bindings/dda_model/optimizers.py

```diff
@@ -0,0 +1,23 @@
+import numpy as np
+
+class AdamOptimizer:
```

Comment:
Nice implementation! A few comments:

  1. Let's write some tests for this - probably `python_tests/test_optimizers.py`
  2. We might think about including the step size / update rate / epsilon / whatever we wanna call it in the optimizer API. All major optimization frameworks do this. For example Pytorch: https://pytorch.org/docs/stable/optim.html, tensorflow/keras: https://keras.io/api/optimizers/, Jax https://jax.readthedocs.io/en/latest/jax.example_libraries.optimizers.html, etc. All of these have roughly the same API (construct optimizer object, call optimizer object with the gradient, current objective value, whatever to return a parameter update).

Reply:
I will get to this later and leave this unresolved. thank you for posting the resources. I am unsure what you mean by 'including the step size' by the way

Reply:
By "including the step size" I mean that the optimizer class can keep track of the step size internally. Right now you are calling the optimizer to get the update, and then scaling that update by a step size. The alternative API (the one used by most ML platforms) is for the optimizer to keep track of the step size inside, and return the scaled update.

Reply:
Okay so have step size as an internal variable, when the optimizer is created?

Reply:
Yes, that's the way that most APIs do it. I am still not sure if it makes more sense for your application to handle the step size inside the optimizer or not, though. The pros of "inside" are that it matches 90% of the other frameworks, so maybe easier to translate or apply optimizers. And it lets you mess with things like a learning rate scheduler more easily (you can have different optimizer classes with different schedulers, for example). The con is that maybe it's too much overhead and you can get away with having the optimization loop manage the step size by itself. I think there may also be optimizers where you need to know the internal variables (like momentum or whatever) to figure out the step size, though I'm not sure.

Reply:
Alright, I wrote down a note to do this after merging the branches